### PR TITLE
rowexec: use opaque keys to identify looked-up rows in lookup joins

### PR DESCRIPTION
**Pull request by yuzefovich**
**sql: add span IDs to fetchers**

This commit implements the idea of Jordan's
78223b921c2ec073667cf2baf920636f24bd2900 to add opaque keys to fetchers.
This allows the fetchers' users to identify the input `roachpb.Span`
that a particular kv was fetched for. I decided to call these keys as
"span IDs" since it seems less ambiguous in general, and especially so
in the streamer world where too many "keys" are flying around already.

Some additional improvements:
- extract the return arguments of a function call into a helper struct
- remove an allocation for `GetResponse`s in the non-streamer code path.

Release note: None

**rowexec: use opaque keys to identify looked-up rows in lookup joins**

This commit refactors the span generators of the lookup joins to use the
span IDs ("opaque keys") for matching between the looked-up rows and the
input rows that correspond to them. Previously, we were maintaining
a map from the span key (bytes value) and had to construct a partial key
for each looked up row to get the matching input row indices. This
commit makes it so that we now assign each span to lookup a unique span
ID that is later used as an index into a two-dimensional slice of ints
to get the matching indices. We still keep a map from the span key for
the de-duplication purposes.

This change is beneficial in several ways:
- removes the need to construct the partial key for each looked up row
- replaces a map lookup based on the partial key into a slice access
- removes the need for the binary search among lookup spans in
range-based lookup joins.

This results in noticeable performance improvements in
[microbenchmarks](https://gist.github.com/yuzefovich/3e5316e0a5ff32fd1b591b0583d78ced).

Here is the difference on all TPCH queries (average over 20 runs of
`tpchvec/perf` roachtest, negative number indicates reduction in
latency):
```
Q2:   0.20s  0.20s  0.49%
Q3:   1.96s  1.66s -14.86%
Q4:   1.67s  1.62s -3.14%
Q5:   2.82s  2.55s -9.55%
Q7:   6.48s  6.93s  6.82%
Q8:   1.08s  0.99s -8.34%
Q9:   7.08s  6.36s -10.13%
Q10:  2.16s  1.94s -9.99%
Q11:  0.55s  0.51s -6.81%
Q12:  8.78s  8.59s -2.13%
Q16:  0.84s  0.82s -2.68%
Q17:  0.49s  0.46s -6.80%
Q20: 14.34s 13.43s -6.36%
Q21:  9.95s  9.28s -6.76%
Q22:  0.58s  0.54s -6.93%
```
(Queries that don't use lookup joins were ignored.)

However, the main motivation behind this change was unblocking (or at
least making it easier) to support lookup joins by the streamer. There,
we want the streamer to operate with these "opaque keys" as well as take
over the span de-duplication logic.

Release note: None

### CODE REVIEW

**Comment by yuzefovich** [comment]
Preliminary benchmark numbers show [some improvement](https://gist.github.com/yuzefovich/51b4045940c5e046d52e938bb1579fef).

**Comment by yuzefovich** [comment]
I'm currently re-running the microbenchmarks and will update the link accordingly, but I doubt anything will change, so I think it is RFAL. cc @jordanlewis if you're curious to take a look.

**Review by rharding6373 — APPROVED** [review]
The preliminary microbenchmark results look great! Nice improvement! :lgtm: with some minor nits and questions.

> Reviewed 16 of 16 files at r1, 10 of 10 files at r2.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/80172#-:-N0WwNAs7S8WEMBzkTzv:b-zvanl)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @cucaroach and @yuzefovich)

___
*[`pkg/sql/row/fetcher.go` line 53 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/80172#-N0WmLqG2QcDkO3g_9R7:-N0WmLqG2QcDkO3g_9R8:b6z2nzg) ([raw file](https://github.com/cockroachdb/cockroach/blob/c1af1fcc401ff64fc53975d368dc13650edf6ec9/pkg/sql/row/fetcher.go#L53)):*
> ```Go
> // kvBatchFetcherResponse contains a response from the KVBatchFetcher.
> type kvBatchFetcherResponse struct {
> 	// moreKVs is true if there are more keys to fetch in the scan.
> ```

For clarification: is this true if there are more KVs not returned in this response? Or if this response includes KVs?
___
*[`pkg/sql/row/fetcher.go` line 58 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/80172#-N0Wi4vT-VARJH3JMYaD:-N0Wi4vT-VARJH3JMYaE:b-aiwskt) ([raw file](https://github.com/cockroachdb/cockroach/blob/c1af1fcc401ff64fc53975d368dc13650edf6ec9/pkg/sql/row/fetcher.go#L58)):*
> ```Go
> 	// the server version. Both must be handled by calling code.
> 	// kvs, if set, is a slice of roachpb.KeyValue, the deserialized kv pairs that
> 	// were fetched.
> ```

These results can be in any order compared to the request, right?
___
*[`pkg/sql/row/fetcher.go` line 366 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/80172#-N0WiLDJ6DjG7metnxNX:-N0WiLDJ6DjG7metnxNY:b-eh1xdk) ([raw file](https://github.com/cockroachdb/cockroach/blob/c1af1fcc401ff64fc53975d368dc13650edf6ec9/pkg/sql/row/fetcher.go#L366)):*
> ```Go
> 	txn *kv.Txn,
> 	spans roachpb.Spans,
> 	spanIDs []int,
> ```

nit: add something about `spanIDs` to the function comment, including what a `nil` input means.
___
*[`pkg/sql/row/kv_batch_fetcher.go` line 333 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/80172#-N0WjtDOC6TxrPyHdmtv:-N0WjtDOC6TxrPyHdmtw:b1jgdur) ([raw file](https://github.com/cockroachdb/cockroach/blob/c1af1fcc401ff64fc53975d368dc13650edf6ec9/pkg/sql/row/kv_batch_fetcher.go#L333)):*
> ```Go
> 			i, j = 0, len(args.spanIDs)-1
> 			for i < j {
> 				f.spanIDs[i], f.spanIDs[j] = f.spanIDs[j], f.spanIDs[i]
> ```

Since spans and spanIDs are the same length, this could be moved into the loop above.
___
*[`pkg/sql/rowexec/joinreader_span_generator.go` line 125 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/80172#-N0Wvrug7xXbxE8VQqvK:-N0Wvrug7xXbxE8VQqvL:bepmnt8) ([raw file](https://github.com/cockroachdb/cockroach/blob/973b666725b8b2690be9c37012f5380dfeba2fff/pkg/sql/rowexec/joinreader_span_generator.go#L125)):*
> ```Go
> // memUsage returns the size of the data structures in the spanIDHelper for
> // memory accounting purposes.
> func (h *spanIDHelper) memUsage() int64 {
> ```

Is this currently used anywhere?

**Review by yuzefovich — COMMENTED** [review]
Updated the microbenchmark numbers in the gist. There appear to be some regressions, but I'm inclined to not look too closely at them since rerunning some subsets ([onetoone](https://gist.github.com/yuzefovich/59fab3bfa8e2ff677ae2add081c54fc8), [onetofour](https://gist.github.com/yuzefovich/9b0636f6e7a27e16f1bc4ae9dd06908c)) of benchmarks on exactly the same SHAs produces different results.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/80172#-:-N0X_1kH0XMDAALryDkH:bb7tqln)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 1 stale) (waiting on @cucaroach and @rharding6373)

___
*[`pkg/sql/row/fetcher.go` line 53 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/80172#-N0WmLqG2QcDkO3g_9R7:-N0XUzBx8qWjMQitDqb1:btoxxlb) ([raw file](https://github.com/cockroachdb/cockroach/blob/c1af1fcc401ff64fc53975d368dc13650edf6ec9/pkg/sql/row/fetcher.go#L53)):*
<details><summary><i>Previously, rharding6373 (Rachael Harding) wrote…</i></summary><blockquote>

For clarification: is this true if there are more KVs not returned in this response? Or if this response includes KVs?
</blockquote></details>

Reworded, let me know if it is more clear.
___
*[`pkg/sql/row/fetcher.go` line 58 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/80172#-N0Wi4vT-VARJH3JMYaD:-N0XV9gc-MKimHCJSjOv:bpj1p82) ([raw file](https://github.com/cockroachdb/cockroach/blob/c1af1fcc401ff64fc53975d368dc13650edf6ec9/pkg/sql/row/fetcher.go#L58)):*
<details><summary><i>Previously, rharding6373 (Rachael Harding) wrote…</i></summary><blockquote>

These results can be in any order compared to the request, right?
</blockquote></details>

Hm, when used by SQL, `kvs` is only populated for GetResponses, and it will contain a single KV pair; we have a couple of other `kvBatchFetcher` implementations which just set `kvs` explicitly themselves.

Not sure what more we should say here in the comment, thoughts?
___
*[`pkg/sql/row/kv_batch_fetcher.go` line 333 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/80172#-N0WjtDOC6TxrPyHdmtv:-N0XXT7L20ARmqhZSUU2:b-i34px3) ([raw file](https://github.com/cockroachdb/cockroach/blob/c1af1fcc401ff64fc53975d368dc13650edf6ec9/pkg/sql/row/kv_batch_fetcher.go#L333)):*
<details><summary><i>Previously, rharding6373 (Rachael Harding) wrote…</i></summary><blockquote>

Since spans and spanIDs are the same length, this could be moved into the loop above.
</blockquote></details>

I originally thought about having an extra conditional inside of the loop would be bad for performance, but this code block doesn't seem that performance-sensitive (and branch prediction should work well anyway), so I implemented your suggestion.
___
*[`pkg/sql/rowexec/joinreader_span_generator.go` line 125 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/80172#-N0Wvrug7xXbxE8VQqvK:-N0XYFbaAiCzieNoPjIL:b84l9zd) ([raw file](https://github.com/cockroachdb/cockroach/blob/973b666725b8b2690be9c37012f5380dfeba2fff/pkg/sql/rowexec/joinreader_span_generator.go#L125)):*
<details><summary><i>Previously, rharding6373 (Rachael Harding) wrote…</i></summary><blockquote>

Is this currently used anywhere?
</blockquote></details>

Yes - since the span generators embed this struct, this method is called "transitively" in `generateSpans`.

**Review by cucaroach — COMMENTED** [review]
Wow, this is nice!  I took a cursory look and will look at it more closely tomorrow with some fresh eyes.

> Reviewed 7 of 16 files at r1, 12 of 12 files at r3, 4 of 10 files at r4.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/80172#-:-N0Y0sju8eFL6_mHrl8r:b-2od8yu)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 1 stale) (waiting on @cucaroach and @rharding6373)

**Comment by jordanlewis** [comment]
Exciting improvement to the microbenchmarks!

**Review by mgartner — APPROVED** [review]
Very cool!

> Reviewed 7 of 16 files at r1, 12 of 12 files at r3, 10 of 10 files at r4, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/80172#-:-N0fR2BZ33dIQ4VcMlbr:b-jdo7t9)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 1 stale) (waiting on @rharding6373 and @yuzefovich)

___
*[`pkg/sql/row/fetcher.go` line 53 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/80172#-N0WmLqG2QcDkO3g_9R7:-N0fJbs65I-fR6UE5YXx:b-o8n4dx) ([raw file](https://github.com/cockroachdb/cockroach/blob/c1af1fcc401ff64fc53975d368dc13650edf6ec9/pkg/sql/row/fetcher.go#L53)):*
<details><summary><i>Previously, yuzefovich (Yahor Yuzefovich) wrote…</i></summary><blockquote>

Reworded, let me know if it is more clear.
</blockquote></details>

So `moreKVs` is `false` iff `kvs` is empty, correct? In that case, should we make this a method the returns `len(kvs) == 0` instead of a boolean?

nit: maybe it'd be more clear to use the inverse and rename to something like `finalFetch`, but I don't feel too strongly about this.
___
*[`pkg/sql/row/kv_batch_fetcher.go` line 78 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/80172#-N0fNZPx7C83gTUHVoGM:-N0fNZPx7C83gTUHVoGN:b1snm9c) ([raw file](https://github.com/cockroachdb/cockroach/blob/2dfbcdb6c4c7b9b1d4d7af65b2af023fa0289a06/pkg/sql/row/kv_batch_fetcher.go#L78)):*
> ```Go
> 	// corresponding spans in the spans slice. In other words, it contains
> 	// 1-to-1 mapping with the spans slice allowing the caller to easily
> 	// identify the input span that produced a particular kv.
> ```

I think this could be more clear that the caller can easily identify the input span the produced a particular KV because the span ID is include in the kvBatchFetcherResponse.

What's the reason that the span's ordinal is insufficient for this? Is it because the `if args.reverse { }` reversal would mess this up?

Did you consider creating a struct that combines this with `roachpb.Spans` or `roachpb.Spans`, like:

```go
type identifiableSpans {
  spans roachpb.Spans
  spanIDs []int
}
```
___
*[`pkg/sql/row/kv_fetcher.go` line 57 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/80172#-N0fOpSs71HrAcljkFpa:-N0fOpSs71HrAcljkFpb:brq8ddt) ([raw file](https://github.com/cockroachdb/cockroach/blob/2dfbcdb6c4c7b9b1d4d7af65b2af023fa0289a06/pkg/sql/row/kv_fetcher.go#L57)):*
> ```Go
> // slice, but it will **not** perform the memory accounting. It is the caller's
> // responsibility to track the memory under the spanIDs slice, and the slice
> // can only be reused once the fetcher has been closed.
> ```

So does this imply an invariant that the fetcher will not increase the cap of `spanIDs`? If so, can you make that explicit in this comment?

**Review by cucaroach — APPROVED** [review]
> Reviewed 5 of 10 files at r4.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/80172)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 1 stale) (waiting on @rharding6373 and @yuzefovich)

___
*[`pkg/sql/rowexec/zigzagjoiner.go` line 648 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/80172#-N0fWEAiChHaek67PjM8:-N0fWEAiChHaek67PjM9:b-vifc61) ([raw file](https://github.com/cockroachdb/cockroach/blob/5c68b6cebd04d12697f0d7578e615bc6cf3719c3/pkg/sql/rowexec/zigzagjoiner.go#L648)):*
> ```Go
> 			txn,
> 			roachpb.Spans{roachpb.Span{Key: curInfo.key, EndKey: curInfo.endKey}},
> 			nil, /* spanIDs */
> ```

nit: shouldn't the comment go inside the comma?
___
*[`pkg/sql/rowexec/zigzagjoiner.go` line 791 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/80172#-N0fWLOf6xpOoxOpPUUl:-N0fWLOf6xpOoxOpPUUm:b-7ropln) ([raw file](https://github.com/cockroachdb/cockroach/blob/5c68b6cebd04d12697f0d7578e615bc6cf3719c3/pkg/sql/rowexec/zigzagjoiner.go#L791)):*
> ```Go
> 			z.FlowCtx.Txn,
> 			roachpb.Spans{roachpb.Span{Key: curInfo.key, EndKey: curInfo.endKey}},
> 			nil, /* spanIDs */
> ```

ditto

**Review by yuzefovich — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/80172)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 1 stale) (waiting on @cucaroach, @mgartner, and @rharding6373)

___
*[`pkg/sql/row/fetcher.go` line 53 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/80172#-N0WmLqG2QcDkO3g_9R7:-N0fhF91DfdHrZwJiyUo:bqtwu4t) ([raw file](https://github.com/cockroachdb/cockroach/blob/c1af1fcc401ff64fc53975d368dc13650edf6ec9/pkg/sql/row/fetcher.go#L53)):*
<details><summary><i>Previously, mgartner (Marcus Gartner) wrote…</i></summary><blockquote>

> a method the returns len(kvs) == 0 instead of a boolean

Meant to say: "a method that returns `len(kvs) == 0` instead of a boolean field".
</blockquote></details>

`moreKVs` is `false` iff `len(kvs) == 0 && len(batchResponse) == 0`. I removed this boolean altogether since it seems to be confusing.
___
*[`pkg/sql/row/kv_batch_fetcher.go` line 78 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/80172#-N0fNZPx7C83gTUHVoGM:-N0flHnf1iX9hPMJzIgq:bsmrod5) ([raw file](https://github.com/cockroachdb/cockroach/blob/2dfbcdb6c4c7b9b1d4d7af65b2af023fa0289a06/pkg/sql/row/kv_batch_fetcher.go#L78)):*
Reworded.

> What's the reason that the span's ordinal is insufficient for this?

In the case of lookup joins, for example, it is possible to have multiple spans correspond to a single span ID - this will be the case when we can take a generated span and split it into several column-family-specific spans. This example forces us to have "user-defined span IDs" (because the lookup join wants to mark multiple spans given to the fetcher with the same span ID), so the fetcher has to play along.

> Did you consider creating a struct that combines this with `roachpb.Spans`?

Do you mean to use this new struct only inside of `txnKVFetcher` (to make the link between `spans` and `spanIDs` more explicit) or ask the caller to provide the struct as an argument? The latter seems redundant since most callers won't use these spanIDs. The former doesn't seem to make things cleaner to me, but I don't have a strong preference here.
___
*[`pkg/sql/row/kv_fetcher.go` line 57 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/80172#-N0fOpSs71HrAcljkFpa:-N0fjRfF2LPovPeJ3ORe:b-9f087b) ([raw file](https://github.com/cockroachdb/cockroach/blob/2dfbcdb6c4c7b9b1d4d7af65b2af023fa0289a06/pkg/sql/row/kv_fetcher.go#L57)):*
<details><summary><i>Previously, mgartner (Marcus Gartner) wrote…</i></summary><blockquote>

So does this imply an invariant that the fetcher will not increase the cap of `spanIDs`? If so, can you make that explicit in this comment?
</blockquote></details>

Yep, done.
___
*[`pkg/sql/rowexec/zigzagjoiner.go` line 648 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/80172#-N0fWEAiChHaek67PjM8:-N0fj_9w3ezSi3v4WRFH:b-1kpgu0) ([raw file](https://github.com/cockroachdb/cockroach/blob/5c68b6cebd04d12697f0d7578e615bc6cf3719c3/pkg/sql/rowexec/zigzagjoiner.go#L648)):*
<details><summary><i>Previously, cucaroach (Tommy Reilly) wrote…</i></summary><blockquote>

nit: shouldn't the comment go inside the comma?
</blockquote></details>

We have no control over it - this is done automatically by the formatter.
___
*[`pkg/sql/rowexec/zigzagjoiner.go` line 791 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/80172#-N0fWLOf6xpOoxOpPUUl:-N0fjhmHE1gv-soHOeIo:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/5c68b6cebd04d12697f0d7578e615bc6cf3719c3/pkg/sql/rowexec/zigzagjoiner.go#L791)):*
<details><summary><i>Previously, cucaroach (Tommy Reilly) wrote…</i></summary><blockquote>

ditto
</blockquote></details>

Done.

**Review by yuzefovich — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/80172)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 1 stale) (waiting on @cucaroach, @mgartner, and @rharding6373)

___
*[`pkg/sql/row/fetcher.go` line 53 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/80172#-N0WmLqG2QcDkO3g_9R7:-N0frY0bBqym0gGOzxG8:b-w6m9c3) ([raw file](https://github.com/cockroachdb/cockroach/blob/c1af1fcc401ff64fc53975d368dc13650edf6ec9/pkg/sql/row/fetcher.go#L53)):*
<details><summary><i>Previously, yuzefovich (Yahor Yuzefovich) wrote…</i></summary><blockquote>

`moreKVs` is `false` iff `len(kvs) == 0 && len(batchResponse) == 0`. I removed this boolean altogether since it seems to be confusing.
</blockquote></details>

Hm, looks like CI is unhappy and simply removing the boolean is not that simple.

**Review by yuzefovich — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/80172)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 1 stale) (waiting on @cucaroach, @mgartner, and @rharding6373)

___
*[`pkg/sql/row/fetcher.go` line 53 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/80172#-N0WmLqG2QcDkO3g_9R7:-N0gAdmKC4SCo4Bzjj_6:b-c2tv7m) ([raw file](https://github.com/cockroachdb/cockroach/blob/c1af1fcc401ff64fc53975d368dc13650edf6ec9/pkg/sql/row/fetcher.go#L53)):*
<details><summary><i>Previously, yuzefovich (Yahor Yuzefovich) wrote…</i></summary><blockquote>

Hm, looks like CI is unhappy and simply removing the boolean is not that simple.
</blockquote></details>

Ok, brought back the boolean. The problem is that an empty response to a ScanRequest might get a `nil` `batchResponse` value which the KVFetcher skips over it, with more data to come later.

**Review by mgartner — COMMENTED** [review]
> Reviewed 1 of 12 files at r5.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/80172)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 1 stale) (waiting on @cucaroach, @mgartner, @rharding6373, and @yuzefovich)

___
*[`pkg/sql/row/fetcher.go` line 53 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/80172#-N0WmLqG2QcDkO3g_9R7:-N0gE47UAapFjQxBnifT:bcpdmhq) ([raw file](https://github.com/cockroachdb/cockroach/blob/c1af1fcc401ff64fc53975d368dc13650edf6ec9/pkg/sql/row/fetcher.go#L53)):*
<details><summary><i>Previously, yuzefovich (Yahor Yuzefovich) wrote…</i></summary><blockquote>

Ok, brought back the boolean. The problem is that an empty response to a ScanRequest might get a `nil` `batchResponse` value which the KVFetcher skips over it, with more data to come later.
</blockquote></details>

Ahh I see. We may have exhausted all the results for a key, but there might be more keys to fetch. :+1: 
___
*[`pkg/sql/row/kv_batch_fetcher.go` line 78 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/80172#-N0fNZPx7C83gTUHVoGM:-N0gENCsEFUJj6A00d2k:bb3z4qy) ([raw file](https://github.com/cockroachdb/cockroach/blob/2dfbcdb6c4c7b9b1d4d7af65b2af023fa0289a06/pkg/sql/row/kv_batch_fetcher.go#L78)):*
> In the case of lookup joins, for example, it is possible to have multiple spans correspond to a single span ID...

Makes sense, thanks for explaining.

> Do you mean to use this new struct only inside of txnKVFetcher (to make the link between spans and spanIDs more explicit) or ask the caller to provide the struct as an argument? The latter seems redundant since most callers won't use these spanIDs. The former doesn't seem to make things cleaner to me, but I don't have a strong preference here.

I was thinking the latter, mostly as a way to encapsulate spans-with-IDs into a single concept. I realized you already did a bit of this in the second commit to implement the sort interface. It's not always true, but in general I think encapsulating tightly-coupled fields into a struct with methods to access and manipulate the struct is generally easier to understand and less error-prone than manipulating multiple fields independently.

Anyway, don't let this block the merge.

**Review by yuzefovich — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/80172)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 1 stale) (waiting on @cucaroach, @mgartner, and @rharding6373)

___
*[`pkg/sql/row/kv_batch_fetcher.go` line 78 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/80172#-N0fNZPx7C83gTUHVoGM:-N0gPuQQ9cAjgIEORF7g:b-gd3hhy) ([raw file](https://github.com/cockroachdb/cockroach/blob/2dfbcdb6c4c7b9b1d4d7af65b2af023fa0289a06/pkg/sql/row/kv_batch_fetcher.go#L78)):*
<details><summary><i>Previously, mgartner (Marcus Gartner) wrote…</i></summary><blockquote>

> In the case of lookup joins, for example, it is possible to have multiple spans correspond to a single span ID...

Makes sense, thanks for explaining.

> Do you mean to use this new struct only inside of txnKVFetcher (to make the link between spans and spanIDs more explicit) or ask the caller to provide the struct as an argument? The latter seems redundant since most callers won't use these spanIDs. The former doesn't seem to make things cleaner to me, but I don't have a strong preference here.

I was thinking the latter, mostly as a way to encapsulate spans-with-IDs into a single concept. I realized you already did a bit of this in the second commit to implement the sort interface. It's not always true, but in general I think encapsulating tightly-coupled fields into a struct with methods to access and manipulate the struct is generally easier to understand and less error-prone than manipulating multiple fields independently.

Anyway, don't let this block the merge.

</blockquote></details>

Added a third commit to implement the idea. I think I agree - it seems less error-prone overall. I intend to squash it into the first commit, but for now it is a separate third commit, PTAL.

**Review by yuzefovich — COMMENTED** [review]
TFTRs!

bors r+

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/80172#-:-N0ggTX-2Ea_i-k7wA8w:b-tlx7ko)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 1 stale) (waiting on @cucaroach, @mgartner, and @rharding6373)

___
*[`pkg/sql/row/kv_batch_fetcher.go` line 78 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/80172#-N0fNZPx7C83gTUHVoGM:-N0ggL244kfGo_i2murG:b-wfipms) ([raw file](https://github.com/cockroachdb/cockroach/blob/2dfbcdb6c4c7b9b1d4d7af65b2af023fa0289a06/pkg/sql/row/kv_batch_fetcher.go#L78)):*
<details><summary><i>Previously, yuzefovich (Yahor Yuzefovich) wrote…</i></summary><blockquote>

Added a third commit to implement the idea. I think I agree - it seems less error-prone overall. I intend to squash it into the first commit, but for now it is a separate third commit, PTAL.
</blockquote></details>

Actually, I think that change won't be controversial, so I just squashed it into the first commit.

I thought a bit about merging the two structs for this "spans-with-spanIDs" concept into one but decided against it since I don't want to export it (they need to be used in two different packages) and navigating the code seems easier with two separate structs. Very little code duplication that we get seems acceptable.

**Comment by mgartner** [comment]
*[`pkg/sql/row/kv_batch_fetcher.go` line 78 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/80172#-N0fNZPx7C83gTUHVoGM:-N0lwmub4W2NMJ9oYPz7:ban9rkp) ([raw file](https://github.com/cockroachdb/cockroach/blob/2dfbcdb6c4c7b9b1d4d7af65b2af023fa0289a06/pkg/sql/row/kv_batch_fetcher.go#L78)):*
<details><summary><i>Previously, yuzefovich (Yahor Yuzefovich) wrote…</i></summary><blockquote>

Actually, I think that change won't be controversial, so I just squashed it into the first commit.

I thought a bit about merging the two structs for this "spans-with-spanIDs" concept into one but decided against it since I don't want to export it (they need to be used in two different packages) and navigating the code seems easier with two separate structs. Very little code duplication that we get seems acceptable.
</blockquote></details>

Sorry for the late drive-by, but this looks great. Another step further would be to export the struct and have the caller create it with or without span IDs. And yet another step further would be to put it in a separate package used by both packages you mention and eliminate the duplicate code. I'll leave it up to you to decide if you think any of that is worth it.

**Pull request merged**
